Source/DafnyStandardLibraries/src/Std/Actions/Actions.dfy

```diff
@@ -0,0 +1,620 @@
+/*******************************************************************************
+ *  Copyright by the contributors to the Dafny Project
+ *  SPDX-License-Identifier: MIT
+ *******************************************************************************/
+
+module Std.Actions {
+
+  import opened Wrappers
+  import opened Frames
+  import opened GenericActions
+  import opened Termination
+  import opened DynamicArray
+  import opened Math
+  import Collections.Seq
+
+  // A composable imperative action.
+  //
+  // Specializes GenericAction to assume its behavior can be specified
+  // independently from most other actions,
+  // and therefore fits into the Valid()/Repr idiom
+  // (and hence extends the Validatable trait).
+  // Its specified behavior is allowed to depend on only
+  // what inputs it consumed and outputs it produced in the past.
+  // 
+  // A key design point for making this possible in Dafny:
+  // the CanConsume and CanProduce predicates,
+  // which the action's specification of behavior are drawn from,
+  // specifically avoid reading the current state of the action.
+  // That is so extrisnic properties of an action do NOT depend on their current state.
+  // This is key to ensure that you can prove properties of a given action that
+  // will continue to hold as the Dafny heap changes.
+  // This approach works because Dafny understands that for a given object,
+  // the implementation of CanConsume/CanProduce cannot change over time.
+  //
+  // The downside is that these are then forced to use quantifiers
+  // to talk about all possible states of an action.
+  // The solution is the trait proof pattern,
+  // where a trait is passed around with an abstract lemma
+  // that can be invoked on the otherwise quantified state as needed.
+  // TODO: see (somewhere I can talk about that pattern more generally)
+  //
+  // This trait is intended to be applicable for any imperative action
+  // regardless of how many input or output values it consumes and produces,
+  // despite only defining two type parameters.
+  // Implementors should feel free to use the () unit type or tuple types
+  // for T and R, under the assumption that
+  // future Dafny backends will be able to easily optimize
+  // away the overhead of passing around a pointless () value
+  // or wrapping up multiple values into a tuple.
+  //
+  // TODO: Better type parameter names for T and R?
+  //
+  // === Errors ===
+  //
+  // Because the Action trait is so general,
+  // there are many error producing and handling patterns that
+  // can be expressed, even within the same type signature:
+  //
+  // 1. An Action<T, Option<R>> can produce None to indicate there is no value,
+  //    but the action could still be called again. Similarly a Result<R, E>
+  //    output could indicate a failure that is only related to that invocation.
```

Comment:
Under this option, is a subsequent invocation of the action obligated/guaranteed to produce "no value" again?

Reply:
Oh, the next bullets answer my question--it depends on what kind of Action it is.